src/client.ts

```diff
@@ -2149,24 +2149,81 @@ export class MatrixClient extends TypedEventEmitter<EmittedEvents, ClientEventHa
     }
 
     /**
-     * Set account data event for the current user.
-     * It will retry the request up to 5 times.
+     * Set account data event for the current user, and wait for the result to be echoed over `/sync`.
+     *
+     * Waiting for the remote echo ensures that a subsequent call to {@link getAccountData} will return the updated
+     * value.
+     *
+     * If called before the client is started with {@link startClient}, logs a warning and falls back to
+     * {@link setAccountDataRaw}.
+     *
+     * Retries the request up to 5 times in the case of an {@link ConnectionError}.
+     *
      * @param eventType - The event type
      * @param content - the contents object for the event
-     * @returns Promise which resolves: an empty object
-     * @returns Rejects: with an error response.
      */
-    public setAccountData<K extends keyof AccountDataEvents>(
+    public async setAccountData<K extends keyof AccountDataEvents>(
+        eventType: K,
+        content: AccountDataEvents[K] | Record<string, never>,
+    ): Promise<EmptyObject> {
+        // If the sync loop is not running, fall back to setAccountDataRaw.
+        if (!this.clientRunning) {
+            logger.warn(
+                "Calling `setAccountData` before the client is started: `getAccountData` may return inconsistent results.",
```

Comment:
`getAccountData` just wouldn't work, it'd return null or whatever its unknown return is, that is unrelated to you calling `setAccountData` so I think this warning is misleading

Reply:
> `getAccountData` just wouldn't work, it'd return null or whatever its unknown return is,

Well that's my point. It won't return the same thing you've just set.

Reply:
yup, that sounds good.

Reply:
At the very least: we should have methods that allow you to wait for the operation to be confirmed.

My first instinct here was to add a new `setAccountDataAndAwaitRemoteEcho` or something, but thanks to the way `ServerSideSecretStorage` is implemented, using it would have meant breaking changes to the library API, so I lost enthusiasm for that.